mountpoint-s3-client/src/mock_client/leaky_bucket.rs

```diff
@@ -0,0 +1,235 @@
+//! A token-based rate limiter.
+//!
+//! This is adapted from https://github.com/Gelbpunkt/leaky-bucket-lite (Apache 2.0) to remove the
+//! Tokio dependency and instead work with the generic async-io and async-lock crates that we use.
```

Comment:
This is an interesting tradeoff. Since `mountpoint-s3-client` already has a dependency on `tokio`, I'm curious about what else led to the decision of reimplementing the leaky bucket instead of using that other crate? Obviously, this now is code we need to review, test, maintain, etc.

Reply:
`mountpoint-s3-client` and `mountpoint-s3` both have dev-dependencies on tokio, but this would require a non-dev dependency (at least when the `mock` feature is enabled). In general we've tried to take an approach of not hardcoding an async runtime, because (for example) the CRT provides its own, and so all of our code should work on any async runtime. So while we could just say that users of this client need to use tokio, I felt like this code is simple and stable enough that it wouldn't hurt to pull it in like this.